### ISSUE TITLE
Selection order of interaction terms in L1 search

### ISSUE THREAD

**Issue by fweber144** [stan-dev/projpred#190]
An L1 search may select interaction terms before the corresponding main terms are selected:
```r
nobsv <- 29
nvars <- 2
set.seed(1235)
x_mat <- matrix(rnorm(nobsv * nvars), nrow = nobsv, ncol = nvars)
x_mat <- cbind(x_mat, x_mat[, 1] * x_mat[, 2])
b <- c(-3, 1, 2)
dat <- data.frame(y = rnorm(nobsv, mean = x_mat %*% b, sd = 1.5), x_mat)
suppressPackageStartupMessages(library(rstanarm))
options(mc.cores = parallel::detectCores(logical = FALSE))
fitobj <- stan_glm(y ~ X1*X2,
                   data = dat,
                   seed = 11403)

library(projpred)
vs <- varsel(fitobj, nclusters = 2, nclusters_pred = 3, seed = 457634)
solution_terms(vs)
```
gives
```
[1] "X1"    "X1:X2" "X2"
```
In contrast, the forward search behaves as expected (only considers an interaction term after all main terms involved in that interaction have been selected). Continuing the example from above:
```r
vs_fw <- varsel(fitobj, method = "forward", nclusters = 2, nclusters_pred = 3,
                seed = 457634)
solution_terms(vs_fw)
```
we get:
```
[1] "X1"    "X2"    "X1:X2"
```
If this issue is not easy to fix, I think it should at least be documented or a warning should be thrown in case of an L1 search with interactions.

**EDIT:** Typo corrected (in the formula).

**Comment by AlejandroCatalina** [comment]
It is not easy to fix unfortunately. We will add documentation.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Frank Weber ***@***.***>
Sent: Thursday, August 12, 2021 11:32:18 AM
To: stan-dev/projpred ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [stan-dev/projpred] Selection order of interaction terms in L1 search (#190)


An L1 search may select interaction terms before the corresponding main terms are selected:

nobsv <- 29
nvars <- 2
set.seed(1235)
x_mat <- matrix(rnorm(nobsv * nvars), nrow = nobsv, ncol = nvars)
x_mat <- cbind(x_mat, x_mat[, 1] * x_mat[, 2])
b <- c(-3, 1, 2)
dat <- data.frame(y = rnorm(nobsv, mean = x_mat %*% b, sd = 1.5), x_mat)
suppressPackageStartupMessages(library(rstanarm))
options(mc.cores = parallel::detectCores(logical = FALSE))
fitobj <- stan_glm(y ~ X1 + X2*X1,
                   data = dat,
                   seed = 11403)

library(projpred)
vs <- varsel(fitobj, nclusters = 2, nclusters_pred = 3, seed = 457634)
solution_terms(vs)

gives

[1] "X1"    "X1:X2" "X2"


In contrast, the forward search behaves as expected (only considers an interaction term after all main terms involved in that interaction have been selected). Continuing the example from above:

vs_fw <- varsel(fitobj, method = "forward", nclusters = 2, nclusters_pred = 3,
                seed = 457634)
solution_terms(vs_fw)

we get:

[1] "X1"    "X2"    "X1:X2"


If this issue is not easy to fix, I think it should at least be documented or a warning should be thrown in case of an L1 search with interactions.

**Comment by fweber144** [comment]
Note: #395 only adds a warning. The underlying issue is still not fixed yet. However, we need to clarify whether we really want to fix that underlying issue (e.g., by modifying the `solution_terms` and `submodls` returned by `search_L1()` in such a case) or not.

**Comment by fweber144** [comment]
Fixed by #420.
